### DISCUSSION TITLE
Padding strategy for DNSCrypt

### DISCUSSION THREAD

**Discussion by msoxzw** [DNSCrypt/dnscrypt-proxy#2139] [Potential issues]
Initially, I found that the "dnscrypt-proxy" consistently sends 1252 bytes UDP packet to the IPv6 DNSCrypt server, e.g. adguard-dns-ipv6. The 1252 bytes, I guess, is based on an MTU of 1280, which is required by the IPv6 specification, minus 20 bytes for the IPv4 header and 8 bytes for the UDP header. However, the size of IPv6 header is 40 bytes, and hence, 1232 bytes UDP packet should be sent on the IPv6 network link.

Furthermore, I am curious about whether DNSCrypt is designed for traffic analysis resistance. The fixed-length or maximal-length padding that DNSCrypt seems to use are not recommended by [RFC 8467](https://datatracker.ietf.org/doc/html/rfc8467).

**Comment by jedisct1** [comment]
As documented in the the relevant [specification](https://github.com/DNSCrypt/dnscrypt-protocol/blob/master/DNSCRYPT-V2-PROTOCOL.txt),  response can be sent over UDP only if its size is less or equal to the question size. This prevents amplification attacks.

`dnscrypt-proxy` uses 512 bytes as a starting point for the question size, and will then gradually adjust this in order to avoid retries or having to use TCP. Padding size of individual queries is randomized, with a maximum overhead of 256 bytes and a block size of 64 bytes. 

The padding of responses use a pseudorandom function to be deterministic for a given nonce. The same question asked multiple times by a genuine client may return responses whose size differs to mitigate correlation, but replay attacks won't allow statistical analysis.

**Comment by msoxzw** [comment]
@jedisct1  Thanks for your detailed explanation.
Moreover, I have confirmed that my IPv6 network can transmit 1444 bytes UDP packet to adguard-dns-ipv6 2a10:50c0::ad1:ff without IP fragmentation (1500 (Standard Ethernet MTU) - 8 (PPPoE header) - 40 (IPv6 header) - 8 (UDP header) = 1444), but still I do not figure out why `dnscrypt-proxy` chooses 1252 bytes rather than 1232 bytes or 1444 bytes. Would you mind enlightening me further?

**Comment by lifenjoiner** [comment]
`DNSCrypt` is not `EDNS`. It is a total encryption of the original query, a new layer, having its own padding strategy. The total length is changing to defeat statistical analysis.
The original query may have its own padding, if it is  `ENDS`.

**Reply by msoxzw** [reply]
The original query that I send is just plain DNS query without EDNS.

**Comment by msoxzw** [comment]
I feel that I find the reason why `dnscrypt-proxy` consistently sends 1252 bytes UDP packet:
https://github.com/DNSCrypt/dnscrypt-proxy/blob/dc2fff05be6a559110136e1bc7f971a528c07ecb/ChangeLog#L283-L285
https://github.com/DNSCrypt/dnscrypt-proxy/blob/dc2fff05be6a559110136e1bc7f971a528c07ecb/dnscrypt-proxy/common.go#L37

Accordingly, in my understanding, `dnscrypt-proxy` may not adjust packet size unless EDNS buffer size from server is larger than 1252 bytes. Unfortunately, since [DNS Flag Day 2020](https://dnsflagday.net/2020/), DNS servers and DNS software have increasingly adopt 1232 bytes EDNS buffer size, so `dnscrypt-proxy` does not further tune DNS packet size.

Additionally, according to the [Cloudflare experiments](https://blog.cloudflare.com/increasing-ipv6-mtu/), the default 1252 bytes UDP packet is probably not suitable for 1280 MTU IPv6 network, e.g. AS6830, AS20825, AS31334, and AS29562. Combined with [DNS XL](https://labs.apnic.net/?p=1380), [DNS 2XL](https://labs.apnic.net/?p=1386), and [DNS Flag Day 2020](https://labs.apnic.net/?p=1390) from APNIC labs, the `dnscrypt-proxy`, I suggest, could use [Path MTU Discovery](https://en.wikipedia.org/wiki/Path_MTU_Discovery) to determine optimal UDP packet size, just like many QUIC implementations, e.g. [MsQuic](https://github.com/microsoft/msquic), [Neqo](https://github.com/mozilla/neqo), [ngtcp2](https://github.com/ngtcp2/ngtcp2), [QUICHE](https://github.com/google/quiche), [quic-go](https://github.com/lucas-clemente/quic-go), [Quinn](https://github.com/djc/quinn), and etc.. Likewise, [some techniques](https://blog.cloudflare.com/accelerating-udp-packet-transmission-for-quic/) used in these QUIC implementations, I believe, could benefit both dnscrypt clients and servers.
